x-pack/filebeat/input/s3/input.go

```diff
@@ -0,0 +1,400 @@
+		awsConfig := defaults.Config()
+		awsCredentials := awssdk.Credentials{
+			AccessKeyID:     config.AccessKeyID,
+			SecretAccessKey: config.SecretAccessKey,
+		}
+		if config.SessionToken != "" {
+			awsCredentials.SessionToken = config.SessionToken
+		}
+
+		awsConfig.Credentials = awssdk.StaticCredentialsProvider{
+			Value: awsCredentials,
+		}
+		return awsConfig, nil
+	}
+
+	// If accessKeyID and secretAccessKey is not given, then load from default config
+	// Please see https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-profiles.html
+	// with more details.
+	if config.ProfileName != "" {
+		return external.LoadDefaultAWSConfig(
+			external.WithSharedConfigProfile(config.ProfileName),
+		)
+	}
+	return external.LoadDefaultAWSConfig()
+}
+
+// Run runs the input
+func (p *Input) Run() {
+	p.logger.Debugf("s3", "Run s3 input with queueURLs: %+v", p.config.QueueURLs)
+	for _, queueURL := range p.config.QueueURLs {
+		regionName, err := getRegionFromQueueURL(queueURL)
+		if err != nil {
+			p.logger.Errorf("failed to get region name from queueURL: %s", queueURL)
+			continue
+		}
+
+		awsConfig := p.awsConfig.Copy()
+		awsConfig.Region = regionName
+		svcSQS := sqs.New(awsConfig)
+		svcS3 := s3.New(awsConfig)
+
+		// update message visibility timeout if it's taking longer than 1/2 of
+		// visibilityTimeout to make sure filebeat can finish reading
+		changeMessageVisibilityInput := &sqs.ChangeMessageVisibilityInput{
+			QueueUrl:          &queueURL,
+			VisibilityTimeout: &visibilityTimeout,
+		}
+
+		// receive messages
+		req := svcSQS.ReceiveMessageRequest(
+			&sqs.ReceiveMessageInput{
+				QueueUrl:              &queueURL,
+				MessageAttributeNames: []string{"All"},
+				MaxNumberOfMessages:   &maxNumberOfMessage,
+				VisibilityTimeout:     &visibilityTimeout,
+				WaitTimeSeconds:       &waitTimeSecond,
+			})
+		output, err := req.Send()
+		if err != nil {
+			p.logger.Error("failed to receive message from SQS:", err)
+			continue
+		}
+
+		// process messages received from sqs
+		if len(output.Messages) > 0 {
+			var wg sync.WaitGroup
+			numMessages := len(output.Messages)
+			wg.Add(numMessages)
+
+			for i := range output.Messages {
+				done := make(chan struct{})
+				// launch goroutine to handle each message from sqs
+				go func(message sqs.Message) {
+					defer wg.Done()
+					defer close(done)
+
+					s3Infos, err := handleMessage(message, p.config.BucketNames)
+					if err != nil {
+						p.logger.Error(err.Error())
+					}
+
+					// read from s3 object and create event for each log line
+					p.readS3Object(svcS3, s3Infos)
+
+					// delete message after events are sent
+					err = deleteMessage(queueURL, *message.ReceiptHandle, svcSQS)
+					if err != nil {
+						p.logger.Error(errors.Wrap(err, "deleteMessages failed"))
+					}
+				}(output.Messages[i])
+
+				select {
+				case <-done:
+				case <-time.After(time.Duration(visibilityTimeout/2) * time.Second):
```

Comment:
@exekias I added this after our chat yesterday about changing message visibility for logs that are taking too long to process. I'm not good add using channels and stuff in go so please review this and let me know what you think. I tested this and it works for me 😬

Reply:
this select is looking good, but it sounds like it should run in another go routine? it's currently blocking the loop until close or timeout happen. Something like this should work:

```
// launch goroutine to handle each message from sqs
go func(message sqs.Message) {..}()
go func() {
  for {
     select {
     }
  }
}()
```

This way the keep alive gorotuine runs at the same time as the `readS3Object` one. Both are guaranteed to finish, either with an error or success.

Reply:
Yeah good point! Thanks for pointing this out! I added the separate go routine in the new commit and tested it 👍 Please let me know!